src/OpenTK.Windowing.GraphicsLibraryFramework/Input/KeyboardState.cs

```diff
@@ -163,67 +139,27 @@ public override bool Equals(object obj)
             return false;
         }
 
-        /// <summary>
-        /// Compares two KeyboardState instances.
-        /// </summary>
-        /// <param name="other">The instance to compare two.</param>
-        /// <returns><c>true</c>, if both instances are equal; <c>false</c> otherwise.</returns>
-        public bool Equals(KeyboardState other)
-        {
-            for (var i = 0; i < NumInts; i++)
-            {
-                unsafe
-                {
-                    if (this._keys[i] != other._keys[i])
-                    {
-                        return false;
-                    }
-                }
-            }
-
-            return true;
-        }
-
-        /// <summary>
-        /// Generates a hashcode for the current instance.
-        /// </summary>
-        /// <returns>
-        /// A <see cref="int" /> representing the hashcode for this instance.
-        /// </returns>
-        public override int GetHashCode()
-        {
-            var hashcode = 0;
-            for (var i = 0; i < NumInts; i++)
-            {
-                unsafe
-                {
-                    hashcode ^= 397 * this._keys[i];
-                }
-            }
-
-            return hashcode;
-        }
-
         /// <inheritdoc />
         public override string ToString()
         {
             var builder = new StringBuilder();
             builder.Append('{');
             var first = true;
 
-            for (var key = (Key)1; key <= Key.LastKey; ++key)
+            string[] keys = Enum.GetNames(typeof(Keys));
+
+            foreach (string key in keys)
             {
-                if (IsKeyDown(key))
+                if (IsKeyDown(Enum.Parse<Keys>(key)))
```

Comment:
If we use an indexed loop with two cached arrays: (of enum strings & enum values) we can avoid this parse.